src/regex.c

```diff
@@ -5,150 +5,155 @@
-    if (t->v.str.lastIndex < 0) t->v.str.lastIndex = 0;
-    if (t->fl.fl.re_flags & SLRE_FLAG_G) {
-      begin = utfnshift(begin, t->v.str.lastIndex);
-    }
-    TRY(regex_check_prog(t));
-    if (!slre_exec(t->v.str.prog, t->fl.fl.re_flags, begin, &sub)) {
+static val_t Regex_exec(struct v7 *v7, val_t this_obj, val_t args) {
+  val_t arr = V7_NULL;
+  if (v7_is_regexp(this_obj) && v7_array_length(v7, args) > 0) {
+    val_t s = to_string(v7, v7_array_at(v7, args, 0));
+    size_t len;
+    struct slre_loot sub;
+    struct slre_cap *ptok = sub.caps;
+    const char *begin = v7_to_string(v7, &s, &len);
+    struct v7_regexp *rp = (struct v7_regexp *)v7_to_pointer(this_obj);
+    int flag_g = slre_get_flags(rp->compiled_regexp) & SLRE_FLAG_G;
+    if (rp->lastIndex < 0) rp->lastIndex = 0;
+    if (flag_g) begin = utfnshift((char *)begin, rp->lastIndex);
+
+    if (!slre_exec(rp->compiled_regexp, 0, begin, begin + len, &sub)) {
       int i;
-      arr = v7_push_new_object(v7);
-      v7_set_class(arr, V7_CLASS_ARRAY);
+      arr = v7_create_array(v7);
+
       for (i = 0; i < sub.num_captures; i++, ptok++)
-        v7_append(v7, arr, v7_mkv(v7, V7_TYPE_STR, ptok->start,
-                                  ptok->end - ptok->start, 1));
-      if (t->fl.fl.re_flags & SLRE_FLAG_G)
-        t->v.str.lastIndex = utfnlen(begin, sub.caps->end - begin);
-      return V7_OK;
-    } else {
-      t->v.str.lastIndex = 0;
-    }
+        v7_array_append(v7, arr, v7_create_string(v7, ptok->start,
+                                                  ptok->end - ptok->start, 1));
+      if (flag_g) rp->lastIndex = utfnlen((char *)begin, sub.caps->end - begin);
+    } else
+      rp->lastIndex = 0;
   }
-  TRY(v7_make_and_push(v7, V7_TYPE_NULL));
-  return V7_OK;
-#undef v7
+  return arr;
 }
 
-V7_PRIVATE enum v7_err Regex_test(struct v7_c_func_arg *cfa) {
-#define v7 (cfa->v7) /* Needed for TRY() macro below */
-  TRY(Regex_exec(cfa));
-
-  v7_push_bool(v7, v7_top_val(v7)->type != V7_TYPE_NULL);
-  return V7_OK;
-#undef v7
+static val_t Regex_test(struct v7 *v7, val_t this_obj, val_t args) {
+  return v7_create_boolean(Regex_exec(v7, this_obj, args) != V7_NULL);
 }
 
-V7_PRIVATE void init_regex(void) {
-  init_standard_constructor(V7_CLASS_REGEXP, Regex_ctor);
-
-  SET_METHOD(s_prototypes[V7_CLASS_REGEXP], "exec", Regex_exec);
-  SET_METHOD(s_prototypes[V7_CLASS_REGEXP], "test", Regex_test);
-
-  SET_PROP_FUNC(s_prototypes[V7_CLASS_REGEXP], "global", Regex_global);
-  SET_PROP_FUNC(s_prototypes[V7_CLASS_REGEXP], "ignoreCase", Regex_ignoreCase);
-  SET_PROP_FUNC(s_prototypes[V7_CLASS_REGEXP], "multiline", Regex_multiline);
-  SET_PROP_FUNC(s_prototypes[V7_CLASS_REGEXP], "source", Regex_source);
-  SET_PROP_FUNC(s_prototypes[V7_CLASS_REGEXP], "lastIndex", Regex_lastIndex);
-
-  SET_RO_PROP_V(s_global, "RegExp", s_constructors[V7_CLASS_REGEXP]);
+V7_PRIVATE void init_regex(struct v7 *v7) {
+  val_t ctor = v7_create_cfunction(Regex_ctor);
```

Comment:
We have some new helpers in place that make sure that the constructor and prototype property attributes are correctly set. See `v7_create_cfunction_ctor`, usage example in `string.c`.